Snowflake.Data.Tests/IntegrationTests/StructuredTypesIT.cs

```diff
@@ -0,0 +1,1471 @@
+                using (var command = connection.CreateCommand())
+                {
+                    EnableStructuredTypes(connection);
+                    var invalidMapSFString = "OBJECT_CONSTRUCT('x', 'y')::OBJECT";
+                    command.CommandText = $"SELECT {invalidMapSFString}";
+                    var reader = (SnowflakeDbDataReader)command.ExecuteReader();
+                    Assert.IsTrue(reader.Read());
+
+                    // act
+                    var thrown = Assert.Throws<SnowflakeDbException>(() => reader.GetMap<string, string>(0));
+
+                    // assert
+                    Assert.AreEqual(SFError.STRUCTURED_TYPE_READ_ERROR.GetAttribute<SFErrorAttr>().errorCode, thrown.ErrorCode);
+                    Assert.That(thrown.Message, Does.Contain("Failed to read structured type when getting a map"));
+                }
+            }
+        }
+
+        private TimeZoneInfo GetTimeZone(SnowflakeDbConnection connection)
+        {
+            using (var command = connection.CreateCommand())
+            {
+                command.CommandText = "show parameters like 'timezone'";
+                var reader = (SnowflakeDbDataReader) command.ExecuteReader();
+                Assert.IsTrue(reader.Read());
+                var timeZoneString = reader.GetString(1);
+                return TimeZoneInfoConverter.FindSystemTimeZoneById(timeZoneString);
+            }
+        }
+
+        private string ToOffsetString(TimeSpan timeSpan)
+        {
+            var offsetString = timeSpan.ToString();
+            var secondsIndex = offsetString.LastIndexOf(":");
+            var offsetWithoutSeconds = offsetString.Substring(0, secondsIndex);
+            return offsetWithoutSeconds.StartsWith("+") || offsetWithoutSeconds.StartsWith("-")
+                ? offsetWithoutSeconds
+                : "+" + offsetWithoutSeconds;
+        }
+
+        private void SetTimePrecision(SnowflakeDbConnection connection, int precision)
+        {
+            using (var command = connection.CreateCommand())
+            {
+                command.CommandText = $"ALTER SESSION SET TIMESTAMP_NTZ_OUTPUT_FORMAT = 'YYYY-MM-DD HH24:MI:SS.FF{precision}'";
+                command.ExecuteNonQuery();
+                command.CommandText = $"ALTER SESSION SET TIMESTAMP_OUTPUT_FORMAT = 'YYYY-MM-DD HH24:MI:SS.FF{precision} TZHTZM'";
+                command.ExecuteNonQuery();
+
+            }
+        }
+
+        private void EnableStructuredTypes(SnowflakeDbConnection connection)
+        {
+            using (var command = connection.CreateCommand())
+            {
+                command.CommandText = "ALTER SESSION SET DOTNET_QUERY_RESULT_FORMAT=JSON";
+                command.ExecuteNonQuery();
+            }
+        }
+
+        private string ConvertNewlinesOnWindows(string text) =>
+            RuntimeInformation.IsOSPlatform(OSPlatform.Windows)
+                ? text.Replace("\n", "\r\n")
+                : text;
+    }
+}
```

Comment:
I would:
- split this suite of tests to separate files for Array, Map and Object separately
- introduce maybe even more test groups ie. StructureTypeToArrayConversions, StructureTypeToArrayEdgeCases 
- maybe add Json to the names of those class tests so when we will introduce more tests or parameterize the tests it will be self explanatory